torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1472,44 +1472,68 @@ void ProcessGroupNCCL::shutdown() {
 
 // NOLINTNEXTLINE(bugprone-exception-escape)
 ProcessGroupNCCL::~ProcessGroupNCCL() {
-  VLOG(2) << logPrefix() << "ProcessGroupNCCL destructor entered.";
-
-  if (!terminateProcessGroup_.load()) {
-    if (rank_ % localDeviceCount_ == 0) {
-      TORCH_WARN_ONCE(
-          "WARNING: process group has NOT been destroyed before we destruct ProcessGroupNCCL. ",
-          "On normal program exit, the application should call destroy_process_group to ",
-          "ensure that any pending NCCL operations have finished in this process. "
-          "In rare cases this process can exit before this point and block the progress of "
-          "another member of the process group. This constraint has always been present, "
-          "but this warning has only been added since PyTorch 2.4");
-    }
-    // If user haven't explicitly destroy/shutdown process group, destructor
-    // needs to do so
-    // Note: we have rewritten `shutdown` to represent the destroy behavior.
-    // Here we route to `abort()` explicitly to maintain the old behavior, until
-    // we fix everything.
+  LOG(INFO) << logPrefix() << "ProcessGroupNCCL destructor entered.";
+  bool cudaContextAlive;
+
+  if (terminateProcessGroup_.load())
+    // `shutdown()` or `abort` already called. Skip the favor of disposing
+    // communicators.
+    goto join_threads;
+
+  // If user haven't explicitly destroy/shutdown process group, destructor
+  // needs to do so
+  // First print warning on first rank of each node
+  if (rank_ % localDeviceCount_ == 0) {
+    TORCH_WARN_ONCE(
+        "WARNING: process group has NOT been destroyed before we destruct ProcessGroupNCCL. ",
+        "On normal program exit, the application should call destroy_process_group to ",
+        "ensure that any pending NCCL operations have finished in this process. "
+        "In rare cases this process can exit before this point and block the progress of "
+        "another member of the process group. This constraint has always been present, "
+        "but this warning has only been added since PyTorch 2.4");
+  }
+
+  // Normally we should avoid making CUDA calls in a destructor because CUDA
+  // context can be exiting. Here we speculate whether the CUDA context is alive
+  // by checking whether the watchdog thread is spawned / alive. If it is, it
```

Comment:
well the watchdog thread definitely does NOT keep the cuda context alive.  which is why we wrap the watchdog thread functions in an exception catcher that catches cuda-driver exceptions, and swallow them on the assumption that main() exited and took the context away before the thread shut down.

Reply:
we should probably avoid this speculation thing and just do the try/catch here too?

Reply:
It's complicated. We can chat later. 
Let me divide this PR into two parts -- the no-CUDA-call in `NCCLComm` destructor should be straightforward.

Reply:
Done, this PR changes `NCCLComm` only.
The next PR also removed the speculation.